site/components/map/ServicesMap.tsx

```diff
@@ -222,120 +224,129 @@ const Map = ({
         setPopupInfo({});
     }, []);
 
-    const selectAllStops = async (evt: SyntheticEvent) => {
-        evt.preventDefault();
+    const addSelectedStopsAndServices = async (includeMarkerData?: boolean) => {
+        const parsed = z.array(stopSchema).safeParse(searched);
+        if (!parsed.success) {
+            stateUpdater({
+                ...state,
+                errors: [
+                    ...state.errors.filter((err) => !Object.keys(servicesConsequenceSchema.shape).includes(err.id)),
+                    ...flattenZodErrors(parsed.error),
+                ],
+            });
+        } else {
+            if (showSelectAllText) {
+                const atcoCodes = includeMarkerData ? markerData.map((marker) => marker.atcoCode) : [];
+
+                const servicesInPolygon = includeMarkerData
+                    ? dataSource
+                        ? await fetchServicesByStops({ atcoCodes, includeRoutes: true, dataSource: dataSource })
+                        : await fetchServicesByStops({ atcoCodes, includeRoutes: true })
+                    : [];
+
+                const servicesStopsInPolygon = servicesInPolygon.flatMap((service) => service.stops);
+                const markerDataInAService = includeMarkerData
+                    ? markerData
+                          .filter((marker) => servicesStopsInPolygon.includes(marker.atcoCode))
+                          .map((marker) => {
+                              const services = servicesInPolygon.filter((service) =>
+                                  service.stops.includes(marker.atcoCode),
+                              );
+                              return {
+                                  ...marker,
+                                  serviceIds: services.length > 0 ? services.map((s) => s.id) : undefined,
+                              };
+                          })
+                    : [];
```

Comment:
The servicesMap function is getting pretty significant, it might be best to break it out while we're at this stage 
and introduce some unit tests to as much of the new business logic if possible, if we can break out little sections it'll make it more readable and the tests will give us confidence that this area of the code is working as expected (for future devs as much as for us!)

i.e. breaking out the functionality here into a `getMarkerDataInAService` function and unit testing that separately.

```
const markerDataInAService = includeMarkerData ? getMarkerDataInAService(markerData, servicesStopsInPolygon, servicesInPolygon) : []

function getMarkerDataInAService(markerData, servicesStopsInPolygon, servicesInPolygon) {
    return markerData
        .filter((marker) => servicesStopsInPolygon.includes(marker.atcoCode))
        .map((marker) => {
            const services = servicesInPolygon.filter((service) =>
                service.stops.includes(marker.atcoCode),
            );
            return {
                ...marker,
                serviceIds: services.length > 0 ? services.map((s) => s.id) : undefined,
            };
        })
}
```

Reply:
done the refactoring for this - anything I should be specifically testing?

Reply:
Awesome, I would create a unit test for getMarkerDataInAService and test all the possible paths i.e.

```
describe('getMarkerDataInAService', () => {
  it('should return an empty array if markerData is an empty array', () => {})
  it('should return an empty array if no matching atcoCode is found within serviceStopsInPolygon', () => {})
  it('should return an array of markers with undefined serviceIds if a matching atcoCode is found in serviceStopsInPolygon but is not found in serviceInPolygon', () => {})
  it('should return markerData within a given service if atcoCode is found in serviceStopsInPolygon and serviceInPolygon', () => {})
})
```

Hope that makes sense but give me a shout if you want to go through it together!

Reply:
done